File: T-Bank/Algorithms-and-Data-Structures-2025/13sprint-Mathematics/contest/D-Последняя_цифра_N!/solution.cpp
#include <iostream>

void solution() {
    /*
    Time Complexity O(N)
    Memory Complexity O(1)
    */
    int n;
    std::scanf("%d", &n);
    if (n == 0 || n == 1) {
        std::printf("1\n");
        return;
    }
    int count2 = 0, count5 = 0;
    int output = 1;
    for (int i = 2; i <= n; ++i) {
        int num = i;
        while (num % 2 == 0) {
            num /= 2;
            ++count2;
        }
        while (num % 5 == 0) {
            num /= 5;
            ++count5;
        }
        output = (output * num) % 10;
    }
    int extra2s = count2 - count5;
    for (int i = 0; i < extra2s; ++i) output = (output * 2) % 10;
    std::printf("%d\n", output);
}

int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/13sprint-Mathematics/contest/D-Последняя_цифра_N!/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(1)
    """
    n: int = int(sys.stdin.readline().rstrip())
    if n <= 1:
        sys.stdout.write('1\n')
        return
    MOD: int = 10
    count2: int = 0
    count5: int = 0
    output: int = 1
    for d in range(2, n + 1):
        num: int = d
        while num & 1 == 0:
            num //= 2
            count2 += 1
        while num % 5 == 0:
            num //= 5
            count5 += 1
        output = (output * num) % MOD
    extra2: int = count2 - count5
    for i in range(extra2): output = (output * 2) % 10
    sys.stdout.write('{}\n'.format(output))
    

if __name__ == '__main__':
    solution()